# NovaTel Customer Retention & Revenue Intelligence System

## Data Cleaning

### Objective

The objective of this notebook is to prepare the datasets for analysis, SQL reporting, KPI development, and dashboard creation.

Cleaning activities focus on:

- Resolving identified data quality issues
- Handling business-driven null values appropriately
- Validating business rules
- Creating analysis-ready datasets
- Preparing data for Power BI modeling

### Inputs

- telecom_customer_churn.csv
- telecom_data_dictionary.csv
- telecom_zipcode_population.csv

### Audit Findings Used

1. Monthly Charge contains 120 negative values.
2. Offer contains unresolved null values.
3. Most other null values were validated as business-driven and should not be treated as errors.

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [2]:
churn_df = pd.read_csv("../data/raw/telecom_customer_churn.csv")

data_dict_df = pd.read_csv(
    "../data/raw/telecom_data_dictionary.csv",
    encoding="latin1"
)

zipcode_df = pd.read_csv(
    "../data/raw/telecom_zipcode_population.csv"
)

In [4]:
churn_df.shape

(7043, 38)

In [5]:
data_dict_df.shape

(40, 3)

In [6]:
zipcode_df.shape

(1671, 2)

## Create Working Copy

### Purpose

Preserve the original dataset and perform all cleaning operations on a separate copy.

In [7]:
df = churn_df.copy()

print(df.shape)

(7043, 38)


## Negative Monthly Charge Investigation

### Purpose

Determine whether negative Monthly Charge values represent data quality issues or valid business scenarios before applying any cleaning rules.

In [8]:
negative_charges = df[df["Monthly Charge"] < 0]

print("Negative Records:", len(negative_charges))

negative_charges["Monthly Charge"].value_counts().sort_index()

Negative Records: 120


Monthly Charge
-10.0    13
-9.0      9
-8.0     12
-7.0     16
-6.0      6
-5.0     11
-4.0     17
-3.0     13
-2.0     10
-1.0     13
Name: count, dtype: int64

### Revenue Consistency Validation

Validate whether customers with negative Monthly Charge values also contain valid revenue information.

In [9]:
negative_charges[
    [
        "Monthly Charge",
        "Total Charges",
        "Total Refunds",
        "Total Long Distance Charges",
        "Total Revenue"
    ]
].describe()

,Monthly Charge,Total Charges,Total Refunds,Total Long Distance Charges,Total Revenue
count,120.000000,120.000000,120.000000,120.000000,120.000000
mean,-5.416667,2028.856250,3.176667,695.348833,2727.278417
std,2.888934,2120.436069,9.855009,801.869312,2719.182437
min,-10.000000,19.400000,0.000000,0.000000,22.540000
25%,-8.000000,343.025000,0.000000,48.190000,460.800000
50%,-5.000000,1325.975000,0.000000,436.720000,1920.660000
75%,-3.000000,3239.512500,0.000000,932.580000,4410.295000
max,-1.000000,7942.150000,49.240000,3211.920000,10830.970000


### Negative Monthly Charge Decision

Investigation showed that customers with negative Monthly Charge values still contain valid revenue information.

Findings:

- Total Charges remain positive.
- Total Revenue remains positive.
- Revenue calculations do not appear corrupted.
- Negative Monthly Charge values appear to represent a valid business scenario within the dataset.

Cleaning Decision:

- No modification applied.
- Values retained in the dataset.

## Offer Null Investigation

### Purpose

Determine whether null values in the Offer field represent a valid business scenario or require imputation.

In [10]:
df["Offer"].value_counts(dropna=False)

Offer
NaN        3877
Offer B     824
Offer E     805
Offer D     602
Offer A     520
Offer C     415
Name: count, dtype: int64

### Offer Findings

- Five valid offer categories were identified:
  - Offer A
  - Offer B
  - Offer C
  - Offer D
  - Offer E
- No invalid offer values were detected.
- 3,877 records contain null values.
- The business meaning of null values is not explicitly documented.
- No transformation has been applied at this stage.

## Cleaning Assessment

### Purpose

Review audit findings and determine whether any fields require modification before analysis.

In [11]:
print("Rows:", len(df))
print("Columns:", len(df.columns))
print("Duplicate Records:", df.duplicated().sum())

Rows: 7043
Columns: 38
Duplicate Records: 0


## Cleaning Findings

The dataset was reviewed using audit results and additional validation checks.

Findings:

- No duplicate records were identified.
- No empty records were identified.
- Missing values were determined to be business-driven in most cases.
- Data types are suitable for analysis.
- Revenue-related fields contain valid values.
- Negative Monthly Charge values appear to represent a valid business scenario and were retained.
- Offer null values were retained because their business meaning is not explicitly documented.

### Cleaning Outcome

No mandatory cleaning transformations were required.

The dataset is considered analysis-ready and can proceed to the Data Modeling phase.